[로또 게임 1단계] - 페어 프로그래밍

chunghyeon-kimm opened the pull request:
## 학습로그
___
##### 1. Enum
  - [열거형](https://rangken.github.io/blog/2015/effective-java-5/)
    - 열거형의 특징과 장점
    - 열거형 생성자와 메서드 작성
___
##### 2. 일급 콜렉션
  - [일급 콜렉션](https://jojoldu.tistory.com/412)
    - 일급 콜렉션의 정의와 장점
___
##### 3. static 사용시 주의점
  - [static 사용시 주의점](https://unabated.tistory.com/entry/%EC%99%9C-%EC%9E%90%EB%B0%94%EC%97%90%EC%84%9C-static%EC%9D%98-%EC%82%AC%EC%9A%A9%EC%9D%84-%EC%A7%80%EC%96%91%ED%95%B4%EC%95%BC-%ED%95%98%EB%8A%94%EA%B0%80)
    - static을 남발하면 프로그램이 객체지향적이지 않고, 메모리 효율성이 떨어지게 된다. 또한 변경에 대한 유연성이 떨어진다.

jiwoo-kimm committed:
docs: initial commit

jiwoo-kimm committed:
docs: 기능 구현 목록 작성

jiwoo-kimm committed:
feat:1000원 미만 입력 시 테스트메서드 작성

jiwoo-kimm committed:
feat: 1000의 배수 아닌 값 테스트메서드 작성

jiwoo-kimm committed:
feat: 금액 입력에 문자열 입력시 예외 처리 테스트 메서드 작성

jiwoo-kimm committed:
refactor: perchasePrice, LottoController 생성

jiwoo-kimm committed:
refactor: 테스트 메서드 리팩토링

jiwoo-kimm committed:
test: 공백 입력 처리 테스트 메서드 추가

jiwoo-kimm committed:
refactor: PurchasePrice -> PurchaseCount 변경

jiwoo-kimm committed:
feat: Lotto 클래스, 테스트클래스 생성

jiwoo-kimm committed:
feat: 로또 구매 개수만큼 생성해서 LottoSet에 저장

jiwoo-kimm committed:
feat: 당첨 번호 클래스 생성

chunghyeon-kimm committed:
test: 숫자 개수 6개 확인 테스트 메서드 작성

chunghyeon-kimm committed:
test: 숫자가 아닌 값 입력 테스트 메서드 작성

chunghyeon-kimm committed:
test: 로또 번호 입력범위 1~45 테스트 메서드 작성

chunghyeon-kimm committed:
feat: 당첨 조건 테스트 메서드 작성

chunghyeon-kimm committed:
feat: 통계 클래스, 통계 테스트 클래스 추가

chunghyeon-kimm committed:
refactor: 패키지 구조 및 클래스 이름 변경

chunghyeon-kimm committed:
test: 보너스 번호 확인 테스트 작성

chunghyeon-kimm committed:
feat: 로또 1 게임의 등수 계산

chunghyeon-kimm committed:
feat: 로또 당첨 개수 구하기

chunghyeon-kimm committed:
refactor: 테스트용 로또 클래스 생성

chunghyeon-kimm committed:
feat: 총 상금액 계산 테스트 메서드 및 클래스 생성

chunghyeon-kimm committed:
feat: Controller, InputView, OutputView 1차 완성

chunghyeon-kimm committed:
refactor: DefaultNumbers 한 번만 생성하도록 수정

chunghyeon-kimm committed:
refactor: Test 메서드 DisplayName 정리

chunghyeon-kimm committed:
refactor: DefaultNumbers 지네릭스 수정, 필요없는 클래스파일 삭제

myangw commented on src/main/java/lotto/domain/LottoNumber.java:
```diff
@@ -0,0 +1,55 @@
+        validate(lottoNumber);
+
+        this.lottoNumber = Integer.parseInt(lottoNumber);
+    }
```
클래스 안에서 lottoNumber가 가져야 할 특성이 명확하게 보이네요 👏

myangw commented on src/main/java/lotto/domain/RandomLotto.java:
```diff
@@ -0,0 +1,28 @@
+import java.util.stream.Collectors;
+
+@Getter
+public class RandomLotto extends Lotto {
```
Lotto객체를 상속받게 하신 결정이 나름대로 재사용되는 부분을 분리하기 위한 선택이었다고 보여집니다.

음 그런데 저라면 여기에서 상속보다는 합성을 할 것 같아요. 이유는
1. 단순히 필드를 재사용하고있고
2. 생성자에서 항상 부모클래스의 필드에만 초기화
3. dto 아닌 객체에서 상속을 쓰는 이유는 기본적으로 같은 행동(메서드) 를 공유하면서 다른 동작을 하기 위함인데 이런 로직이 없음
4. 부모 클래스와 강하게 결합되어 서로의 변경에 영향을 줌 -> 두개의 클래스가 하나의 클래스처럼 동작해 분리에 큰 의미가 없어지고 수가 많아진다면 관리가 복잡함

등등이 있습니다.

보다 자세한 이유는 아래 링크 글들 읽어보시면 좋겠고 추가적인 설명이 필요하시면 말씀주세요..!
(나중에 기술서적 함께 읽을 때 조영호 -'오브젝트' 책 읽으시면 좀더 와닿으실 것 같아요..)
https://incheol-jung.gitbook.io/docs/q-and-a/architecture/undefined-2


https://velog.io/@ljinsk3/%EC%83%81%EC%86%8D%EA%B3%BC-%ED%95%A9%EC%84%B1#%EA%B5%AC%ED%98%84-%EC%83%81%EC%86%8D%EA%B3%BC-%EC%9D%B8%ED%84%B0%ED%8E%98%EC%9D%B4%EC%8A%A4-%EC%83%81%EC%86%8D

chunghyeon-kimm replied:
알려주신 링크들 읽어보고 상속과 합성에 대한 이해를 좀 더 깊이 할 수 있었습니다. 감사합니다.

RandomLotto클래스를 상속에서 합성으로 리팩토링하려고 시도해봤는데
상속을 통한 다형성을 이용한 부분에서 문제가 발생해서 고민끝에 RandomLotto를 생성하는 메서드를
Lotto클래스로 옮기고 RandomLotto클래스는 삭제했습니다.

상속을 통해 확장된 부분이 랜덤로또 생성 메서드 하나뿐이라 상속을 이용한 이득보다 비용이 더 크다고 판단했습니다.

같은 이유로 lottoSet 클래스를 상속한 RandomLottoSet 클래스도 메서드를 lottoSet클래스로 이동시키고 삭제했습니다.

myangw replied:
지난 코멘트는 선생님이 구현해놓으신 로직만 보고(Lotto에서 아무 행동도 하지 않은것만 보고..) 저렇게 달았었는데
다시 보니 합성이 적절하지 않은거같습니다. ㅠㅠ 죄송해요.

이유는 우선 RandomLotto - Lotto 클래스를 생성하신 의미를 생각해봤을 때 RandomLotto가 Lotto와 같은 동작을 하면서 + 랜덤하게 로또를 생성한다는 자세한 구현을 해야하기 때문입니다. (둘다 로또인데, Random으로 생성되는 Lotto를 의도하셨을거같아서요. 마치 아이폰과 아이폰12의 관계..) 윗 코멘트에서 `dto 아닌 객체에서 상속을 쓰는 이유는 기본적으로 같은 행동(메서드) 를 공유하면서 다른 동작을 하기 위함` 이렇게 적어놨는데 이게 RandomLotto와 Lotto의 관계인것 같습니다.

기존처럼 상속 관계를 유지하고, Lotto 객체를 추상클래스로 만들어서 Lotto 객체에서 어떤 행동이 필요한지를 정의하고(로또 번호를 생성한다 등), Lotto를 상속한 클래스들도 그 행동을 하게하면 어떨까 싶습니다.

myangw commented on src/main/java/lotto/domain/RandomLotto.java:
```diff
@@ -0,0 +1,28 @@
+        Collections.shuffle(defaultNumbers);
+        return defaultNumbers.subList(INDEX_LOWER_BOUND, INDEX_UPPER_BOUND)
+                .stream()
+                .map(String::valueOf)
```
LottoNumber 객체를 생성할 때 String으로 굳이 바꿔주어야하는 게 이상합니다. LottoNumber 객체 내의 로직에서 String으로서 동작하는 부분이 없고 String으로 받은걸 다시 Integer로 바꿔서 동작하고 있어요.

LottoNumber클래스 생성자로 int를 받게 하면 어떨까요 ?

chunghyeon-kimm replied:
LottoNumber 클래스 생성자 매개변수로 int를 받도록 수정했습니다.

수정 전 구조는 Scanner로 String을 입력받아서 LottoNumber 객체를 생성할 때
생성자에서 입력받은 String이 숫자로만 이루어져있는지, 그리고 그 숫자가 1에서 45 사이 범위에 해당하는지 확인하는
유효성검사를 진행했습니다.

매개변수로 int를 받도록 수정하면서 String에서 int로 변환할 때 숫자로만 이루어져있는지 확인하는 로직을
LottoNumber클래스에서 진행할 수가 없어서 NumberValidateUtils 클래스에 numericCheck 메서드를 만들고
LottoNumber 생성자를 호출하는 각 부분마다 numericCheck 메서드를 먼저 호출해 유효성검사를 진행하게 만들어봤습니다
(LottoParser, purchaseCount, WinningLotto, inputView)

수정하고 보니
객체의 책임 관점에서 보면 LottoNumber클래스 객체의 유효성검사를 해당 클래스에서 직접 하는 게 맞는 것 아닌가
하는 생각이 들었습니다. 이 부분 어떻게 생각하시는지 궁금합니다.

myangw replied:
유효성 검사를 시키는 주체는 해당 클래스이기 때문에 저는 괜찮다고 봅니다.
그리고 무엇보다 이런 유효성 검사는 string -> int로 바꾸기전마다 해야하는것이니 많은 클래스에서 공통으로 쓰는 것일텐데 이런 자잘한 로직을 클래스마다 두면 중복도 생기고요. 

각 클래스마다 numericCheck를 하고-> parseInt를 수행하는게  중복이 되는데
util 클래스에 parse하는 메서드를 만들어 거기서 parse 직전에 유효성 검사를 하게하는것은 어떨까요?

myangw commented on src/main/java/lotto/controller/LottoController.java:
```diff
@@ -0,0 +1,30 @@
+        WinningLottoInputDTO winningLottoInputDTO = View.getWinningLottoAndBonus();
+        WinningLotto winningLotto = new WinningLotto(
+                winningLottoInputDTO.getWinningLottoNumbers(),
+                winningLottoInputDTO.getWinningLottoBonus());
```
winningLottoInputDTO 자체를 argument로 넘겨도 될거같아요. 둘다 winningLotto를 생성하는데에만 쓰이는데 Controller단에서 get해서 넘기신 이유가 있을까요 ?

chunghyeon-kimm replied:
특별한 이유는 없고 WinningLotto클래스의 생성자에 맞는 값을 넘겨주기 위해 이렇게 코딩했습니다.
WinningLotto클래스에 WinningLottoInputDTO클래스의 매개변수를 받는 생성자를 새로 만들어서
DTO객체 자체를 매개변수로 넘기도록 수정했습니다.

myangw commented on src/main/java/lotto/constant/PrizeCondition.java:
```diff
@@ -0,0 +1,43 @@
+
+    private final int matchNumbersCount;
+    private final boolean isBonus;
+
```
enum활용 너무나 적절하네요 👍

myangw commented on src/main/java/lotto/constant/PrizeMoney.java:
```diff
@@ -0,0 +1,19 @@
+
+@Getter
+public enum PrizeMoney {
+
```
PrizeMoney, PrizeMessage, PrizeCondition enum을 각각 따로 생성해두셨는데

중복된 부분이 여러 군데에 흩어져있어서 나중에 변경이 필요하면 두개 이상의 클래스를 모두 고쳐야하는 상황이 생길 수 있어요.
(ex. 상금 수정 시에 PrizeMoney, PrizeMessage 둘다 고쳐야하는 등)
그리고 PrizeMessage의 경우 다른 정보들을 잘 조합해서 만들 수 있을 듯 한데 한번 수정해보시면 좋겠습니다.

chunghyeon-kimm replied:
네. 변경시 수정해야 할 부분이 여러 곳에 흩어져 있는 구조는 매우 나쁜 구조인데, 잘 짚어주셔서 감사합니다.
말씀해주신대로 PrizeMessage는 PrizeCondition과 PrizeMoney를 조합해서 만들도록 수정했습니다.

myangw replied:
음 지난코드보다 개선이 되었는데 여전히 enum필드명이 3개의 enum클래스에 여러군데 흩어져 있어요. 만약에 FIRST SECOND 와 같은 필드명을 바꾼다고 했을때 3개를 다 변경해야합니다.

중복을 없애기 위해

저는 1개의 enum에서 충분히 다 표현가능하다고 생각합니다. 그리고 메세지의 경우에는 보너스볼 빼고는 다 똑같은 문구가 들어가는데 이것도 중복이라.. 메세지 각각을 필드로 들고있는거보다 더 좋은 방법이 없을지 고려해봐주세요

myangw commented on src/main/java/lotto/domain/LottoStatistics.java:
```diff
@@ -0,0 +1,28 @@
+        this.purchaseCount = purchaseCount;
+    }
+
+    public double calcProfitRate() {
```
요구사항에 축약금지 위반이네요 ㅎㅎ


`객체지향 생활체조 원칙 이라는 키워드로 검색해보고 내용을 숙지한다.`
- **규칙 3: 모든 원시값과 문자열을 포장한다.**
- **규칙 5: 줄여쓰지 않는다(축약 금지)**
- **규칙 8: 일급 콜렉션을 쓴다.**

chunghyeon-kimm replied:
습관을 고치는 게 쉽지 않은 것 같습니다. ㅎㅎㅎ
수정했습니다.

myangw commented on src/main/java/lotto/domain/PurchaseCount.java:
```diff
@@ -0,0 +1,37 @@
+
+    private void validate(String input) {
+        if (!isInteger(input)) {
+            throw new IllegalArgumentException("잘못된 입력입니다");
```
예외 메시지를 구체적으로 써주는 습관을 들이시면 읽는 사람 및 클라이언트에게 좋습니다.


어플리케이션 실행해서 100원이나 숫자 아닌 문자를 입력하면 "잘못된 입력입니다" 라고만 뜨고 실행이 종료가 됩니다.
사용자 입장에서 무엇을 잘못했는지까지 알 수 있어야 할 것 같아요 :)

chunghyeon-kimm replied:
네. 좋은 코멘트 감사합니다. 구체적인 메세지를 알려주도록 수정했습니다.

myangw commented on src/main/java/lotto/domain/WinningLotto.java:
```diff
@@ -0,0 +1,50 @@
+    }
+
+    private int getMatchNumbersCount(Lotto targetLotto) {
+        targetLotto.getLottoNumbers().sort(Comparator.comparingInt(LottoNumber::getLottoNumber));
```
winning lotto의 sort는 생성자에서 이루어지고있는데, 사실 이 match number가 몇개인지를 세는 작업 때문에 sort를 한것이라면 생성자가 아닌 이 메서드 안에서 sort를 해야하는 것이 아닐까 싶습니다

chunghyeon-kimm replied:
네. sort의 목적상 생성자에서 수행하는것 보다 해당 메서드에서 수행하는 게 맞는 것 같습니다.
말씀해주신대로 수정했습니다.

myangw commented on src/main/java/lotto/controller/LottoController.java:
```diff
@@ -0,0 +1,30 @@
+    public static void main(String[] args) {
+        start();
+    }
+
```
controller가 main 메서드와 start메서드를 둘다 갖고 있고 View의 메서드들은 모두 static으로 되어있네요. 어떤 이유로 이런 구조를 만드셨나요 ?

chunghyeon-kimm replied:
이 부분은 페어프로그래밍을 함께 진행한 지우선생님이 작성하신 부분인데,
옆에서 보면서 아~ 저렇게 하시는구나 하고 보기만 했었습니다.

뭔가 의문을 던지기에는 static과 non-static에 대한 제 이해가 모자랐던 것 같습니다.
static메서드는 클래스 생성 시 메모리에 올라가고, 인스턴스 멤버를 사용하지 않는 메서드는 static으로 선언할 수 있다
정도로만 알고 있었습니다.

미향선생님께서 코멘트를 주신 의도가 무엇일지 생각해보면서 검색을 해봤습니다.
https://unabated.tistory.com/entry/%EC%99%9C-%EC%9E%90%EB%B0%94%EC%97%90%EC%84%9C-static%EC%9D%98-%EC%82%AC%EC%9A%A9%EC%9D%84-%EC%A7%80%EC%96%91%ED%95%B4%EC%95%BC-%ED%95%98%EB%8A%94%EA%B0%80
이 글을 읽어보고, static의 단점들을 새롭게 알게 되었습니다.

Controller는 main메서드 하나만 가지도록 수정하고, View클래스는 멤버들을 non-static으로 수정했습니다.

혹시 미흡하거나 수정할 부분 있는지 궁금합니다.

myangw replied:
오 좋은 레퍼런스 찾으셨네요 😃 

view를 갈아끼울수도 있어야하고, 객체지향적 코드를 위해서 코멘트 했었습니다.

음 추가적으로, main메서드는 어플리케이션의 제일 처음 시작점인데요. 
이 메서드 안에서 많은 일들을 하기보다 딱 어플리케이션이 시작할 수 있도록 하는 역할만 담고 나머지는 'main'이 아닌 선생님께서 정의하신 어떤 이름으로 추상화하면  어떨까요.

[메인 메서드 관련해서는 이것도 읽어보세요](https://javacpro.tistory.com/11) 
컨트롤러도 컨트롤러의 역할을 담는 클래스라고 본다면 아무 역할이 없는 클래스에서 메인 메서드를 정의하면 좋겠습니다

myangw reviewed (changes requested):
안녕하세요 충현선생님. 코드리뷰가 많이 늦었습니다 ㅠㅠ 전반적으로 고민의 흔적이 많이 보이는 코드였어요. 몇가지 코멘트를 남겼는데 확인 부탁드려요. 고생많으셨습니다!

chunghyeon-kimm committed:
refactor: 메서드 이름 변경(축약금지 위반)

chunghyeon-kimm committed:
refactor: 예외 메세지 수정

chunghyeon-kimm committed:
refactor: 로또 정렬을 생성자가 아닌 비교 메서드에서 하도록 수정

chunghyeon-kimm committed:
refactor: PrizeMessage enum을 PrizeCondition과 PrizeMoney를 이용해 구성

chunghyeon-kimm committed:
refactor: RandomLotto클래스 삭제. Lotto클래스에 RandomLotto 생성 메서드 추가

chunghyeon-kimm committed:
refactor: RandomLottoSet클래스 삭제. LottoSet클래스에 RandomLottoSet 생성하는 메서드 추가

chunghyeon-kimm committed:
refactor: WinningLotto클래스 Lotto클래스 상속에서 포함으로 변경

chunghyeon-kimm committed:
refactor: LottoNumber클래스 생성자 매개변수 String에서 int로 변경

chunghyeon-kimm committed:
refactor: LottoNumber클래스 생성자 매개변수 String에서 int로 변경 및 관련된 부분 수정

- NumberValidateUtils에 String을 입력받아 숫자로만 구성되어있는지 확인하는 numericCheck메서드 추가
- LottoNumber클래스 객체 생성하는 코드들에 numericCheck메서드 적용(유효성 검사)

chunghyeon-kimm committed:
refactor: WinningLotto클래스에 WinningLottoInputDTO 객체를 매개변수로 받는 생성자 추가

chunghyeon-kimm committed:
refactor: View 클래스 멤버들 static에서 nonstatic으로 변경, LottoController start메서드 제거

myangw commented:
이전에 코멘트 달았던 부분에다가 몇개 새로 추가해서 달았습니다. 확인부탁드려요.

수정은 2단계 구현이랑 같이 해주셔도 될것 같아 우선 1단계는 머지하겠습니다.
